JBPM-9175: Stunner - Usability improvements around selection / drag capabilities

handreyrc opened the pull request:
@romartin, @inodeman: Please, could you review this PR?

See https://issues.redhat.com/browse/JBPM-9175

Assembly with:
* https://github.com/kiegroup/kie-wb-common/pull/3319
* https://github.com/kiegroup/lienzo-core/pull/111
* https://github.com/kiegroup/lienzo-tests/pull/94

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
@romartin, @LuboTerifaj : I've generated the artifacts locally and they are available through the link:
https://drive.google.com/file/d/1fvAtNkcpOj6Hrittwyl_EcWMUGzpTDvu/view?usp=sharing

Related PRs:
https://github.com/kiegroup/lienzo-core/pull/111
https://github.com/kiegroup/lienzo-tests/pull/94

Please, let me know if it suits you. Thanks

handreyrc commented:
Jenkins retest please

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Jenkins execute full downstream build

domhanak commented:
@handreyrc Hi, I am modeling processes using this PR and I am noticing 1-2s delay (2s is worst case) for showing toolbox quickmenu, when selecting nodes. Is that expected? 
Examples:
- Select EndNode and then select HT after, wait for HT quick menu
After some repetition of these steps the issue is gone, but overall feeling when modelling is that the selection is slower and I did not like it as a user. Maybe some1 else could do a UX check from their POV? @romartin

For the issue with dragging - JBPM-8253
There are three cases I tested:

1. When you just click&hold&drag the quick menu of node is not shown and only change is that it will have blue borders when dragged. This changes based on if we had the node selected or we had zero selections on canvas.
If there are no selections - in some cases node is kept with black borders, but is some it is selected with blue
If there is selection on canvas - for the node we drag it is correct, for different node check case 3 further down the comment.

2. When you Select node first  and then click&hold&drag, the quick menu is shown and selected node has blue borders. This is correct and fixes the issue.

3. Select a node A and then start dragging a different node B. Selection is kept on node A and different node B is being dragged, moved and their borders are still black. Is this expected?

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Jenkins execute full downstream build

romartin commented:
Hey @handreyrc , are the issues from @domhanak fixed in this PR? all of them? what's the state for those please?

romartin commented:
Hey @hasys , adding you as for the eng. review on this as well. Check related PRs. Thanks!

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/ConnectionAcceptorControlImpl.java:
```diff
@@ -257,18 +258,38 @@ private boolean isAccept(final CommandResult<CanvasViolation> result) {
                            final boolean valid) {
-        canvasHighlight.unhighLight();
+        // keep selected nodes
+        ensureSelection(node);
```
I'm not following you on these changes about calling `ensureSelection` and about the flag for the `canvasHighlight.unhighLight` call. Which is the goal behind those changes, about usability? about performance?

handreyrc replied:
@romartin This soves the issue where the shape loses decoration when the sequence flows change positions while dragging. There are 4 styles for a shape NONE / SELECTED / HIGHLIHT / INVALID and we can have only one per time. 
When the sequence flows change position while dragging a shape, the acceptors are unaware of the selection state of the shape and set the style to NONE. 
He idea is to tag selected nodes and ensure they remain selected while the acceptor reevaluates connections.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/ConnectionAcceptorControlImpl.java:
```diff
@@ -257,18 +258,38 @@ private boolean isAccept(final CommandResult<CanvasViolation> result) {
+        }
+    }
+
+    private boolean isNodeSelected(final Node node) {
```
well it's true, but I think we should not consider a node being selected by its state for the shape (`ShapeState.SELECTED`). 
For example, at some point it, if the node gets into an invalid state, the shape will have the state as `ShapeState.INVALID`, but it can be selected as well, so forms, toolbox and any other component that queries the actual selection via the canvas controls. 
So I think that we should consider a node selected by looking at the canvas `SelectionControl`, not looking at the visual attributes.
Anyway, look at my comment above, not really sure what's the goal for this, so maybe we've not to rely on selection at some point, let's keep brainstorming here

handreyrc replied:
@romartin The SelectionControl was my first pick to solve the issue. However, I notice that the UUID of the selected shape recovered from the SelectionControl doesn't match the ones in the internal list kept by the CanvasHighlight. Thats why I changed the approach for checking the style. 
Regarding to possible issues with the INVALID state, it only happens when you try to stablish new connections by dragging sequence flows. In case of dragging a shape, the connections already stablished are aways valid. I couldn't detect any issue while testing.

handreyrc replied:
@romartin, I've changed the implementation to use SelectionControl instead of checking for the current style as you suggested. It is a lot cleanner now!
Thanks for the insight!

romartin replied:
Yeah it makes not much sense that the SelectionManager ids do not match with the real selection on the UI.... anyway, please rely on the `SelectionControl` (Stunner), which it's the one that holds the selection state for the UI. Does it works?

handreyrc replied:
hey @romartin, I've changed the approach to solve this issue. The root cause was in the methods allowTarget() and allowSource().

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasHighlight.java:
```diff
@@ -35,6 +35,7 @@
 
     private final Set<String> uuids;
     private AbstractCanvasHandler canvasHandler;
+    private String selectedUUID = "";
```
As also commented above, I need to figure out why you're adding this new `selectedUUID` field, added more state in the class, to achieve the overall goal. 
The lifecycle for this `CanvasHighlight` stuff was supposed to be really simple - by calling the `highLight` method it's just changing the visual style for each node, and finally by calling `unhighLight` it will just get back to normal visualization. And now, by looking at the changes on the public methods, it's a bit more confusing to me the lifecycle.
I'm not complaining about this, you've more feedback than me hehe, just let's try to clarify the goals to achieve and do collective thinking on the best way to do it :)

handreyrc replied:
@romartin The ideia of this field it to tag the selected shape if there is one in the context.
This issue is happening in a very complex scenario, the highlight is called multiple times invloving the same shape when the inbound and outbound of the shape and the sequence flow are evaluated.

handreyrc replied:
hey @romartin, I've changed the approach to solve this issue. The root cause was in the methods allowTarget() and allowSource().

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/util/CanvasHighlight.java:
```diff
@@ -77,16 +84,29 @@ public CanvasHighlight none(final Element<?> node) {
+        return unhighLight(false);
+    }
+
+    public CanvasHighlight unhighLight(boolean keepSelection) {
```
a bit unclear... what does `keepSelection` flag means? which selection?

handreyrc replied:
@romartin There are situations where we want to clean everything and set the style to NONE.
This flag is used for those contexts. If "True" keep the selection if "False" clean everything.

handreyrc replied:
hey @romartin, I've changed the approach to solve this issue.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/shape/impl/ConnectorShape.java:
```diff
@@ -85,4 +85,9 @@ public void applyPosition(final Edge<ViewConnector<W>, Node> element,
     }
+
+    @Override
+    public ShapeState getShapeState() {
```
if we don't rely on the shape state in the connection acceptor maybe there will be no need to propagate all these delegate methods in all changes classes, although is stricly correct doing it as well, let's keep discussing it

handreyrc replied:
@romartin Sure I'm opened to improve this implementation.

handreyrc replied:
@romartin By the use of the SelectionControl it is not necessary anymore. I've already removed these changes.

handreyrc replied:
hey @romartin, I've changed the approach to solve this issue. it is no longer needed.

handreyrc commented:
> Hey @handreyrc , are the issues from @domhanak fixed in this PR? all of them? what's the state for those please?

Hey @romartin,  here goes the status:
1 - It was an issue due to an uncovered condition and it is already fixed.
2-  It is just an ok for the implementation.
3-  It is an unrealted issue. However It was implemented to work this way sinse the selection is triggered on mouse-up. 
Thanks!

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Hey @romartin, I found a different outlook to tackle this selection issue. The highlight  method will be called only when we have changes on source or target nodes. 
Please, could you review it?

handreyrc commented:
Jenkins execute full downstream build

hasys commented:
Hi @handreyrc, overall looks good, thank you! But can I ask you to provide kind of summary what finally was fixed in this PR and what wasn't? I am tried several reproducers and from comments I understood it is fixed, but I still can reproduce it. And some fixed issues I even can't reproduce on master so not sure was it fixed or it worked before. Thank you!

handreyrc commented:
Hey @hasys, here goes the summary of issues fixed. 
I also added some notes below. If you have any problems reproducing the original issues, feel free to contact me directly.

* Toolbox - Remove/reduce animations around showing or hiding the toolbox on selection, the final experience results slow for end users. 
Note:The animation duration was set to 0.

* Lienzo multiple selection drag-and-drop - In pure lienzo, when multiple shapes are selected and then dragged and dropped multiple times by clicking on different shapes in the selection. At least one shape moves to the top of the selection shape causing the deselection of the other shapes. In addition, the group selection is forced to the top by multiple calls to the method moveUp while dragging. These calls it can be optimized to a single call.
Note: This is a pure lienzo issue and is not reproducible in stunner (stunner has it's own implementation on top of it).You will need the my-lienzo-tests (https://github.com/romartin/my-lienzo-tests.git) project to verify this one. 

* Scrollbars moving on click - With a diagram long vertically, scroll to the very bottom of the diagram and then click repeatedly on a blank area around the last coponent in the vertical. It will causes the scrollbar to move alone some pixels. The canvas bounds are being reacalculated and the scrollbars position updated on every click. It can be optimized to updated the scrollbars positions only if there are changes in the canvas bounds.
Note: This issue is more evident when the canvas is presented with reduced size. Open the properties panel, increase the zoom or try to reduce the size of your browser window.

* https://issues.redhat.com/browse/JBPM-8253
Note: I only fixed the highlight issue while dragging a selected shape described in the original issue.

* https://issues.redhat.com/browse/JBPM-9019 

Thanks!

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/controls/toolbox/AbstractToolboxControl.java:
```diff
@@ -52,7 +52,10 @@
-            return (null == id && count == 0) || (null != id && count == 1 && id.equals(s)) || (null != id && count > 1 && !id.equals(s));
+            return (null == id && count == 0) ||
+                    (null != id && count == 1 && id.equals(s)) ||
+                    (null != id && count > 1 && !id.equals(s)) ||
```
```java
(null != id && count > 1 && !id.equals(s)) || (null != id && count > 1 && id.equals(s))
```

can be simplified to 

```java
(null != id && count > 1)
```

handreyrc replied:
Hey @hasys, good catch! 
I've already updated the code.

hasys commented:
Hi @domhanak,

my 2c about your questions:

1. (about 1-2s delays during task activation), I think it is related to Forms and Business Central part, can you reproduce it in Kogito or in Business Central but with closed Properties?
2. (click&hold&drag the quick menu is not shown), it is valid point, but, I guess unrelated to this PR and I suggest to move it to separate issue/PR (see some ideas below)
3. (select a node and drag different one), I think it is a valid point as well but has the same root cause (selection actions should be moved from MouseUp to MouseDown) and it should be a separate issue (see some ideas below)

Some ideas about 2&3 I guess it is done on MouseUp because on MouseDown we don't know is user want to select an area or one specific node, but we can try to add some logic that if MouseDown on the node - it will be one node selection, if on the canvas - it will be a selection. I think this issue is worth investigations, but I afraid it is not so trivial as change MouseUp to MouseDown and will desire some more investigations so I suggest to move to the separate issue.

Thank you!

handreyrc commented:
Hey @hasys, @domhanak, @romartin : 
* I've checked all previous remarks and I made some changes to improve the code quality overall.
* As we agreed, we do have some other issues around these functionalities, however it would be better to report them separately.

Here goes the link to the artifacts for testing it:
https://drive.google.com/file/d/1L4Nh6f-kGDy_MCUkbWhkDeNM0cmsNWlf/view?usp=sharing

Thanks!

handreyrc commented:
Jenkins execute full downstream build

handreyrc commented:
Jenkins retest please

romartin commented:
Jenkins retest please

romartin commented:
please @handreyrc lemme know once the normal CI build is ok, so I'll merge. Thanks!

romartin commented:
Jenkins retest this

handreyrc commented:
Jenkins execute compile downstream build
